### DISCUSSION TITLE
🚀 Brainstorming & Future Roadmap: Meshtastic, Offline Networks, and TinyML

### DISCUSSION THREAD

**Discussion by TN666** [batear-io/batear#1] [Ideas]
# 🚀 Brainstorming & Future Roadmap: Meshtastic, Offline Networks, and TinyML

Welcome everyone! 👋 

Thanks to the recent features on Hackaday and Hacker News, we’ve had a massive influx of amazing ideas from the community. **Batear** is currently a solid, $15 flashable baseline using the Goertzel algorithm, but this is just the beginning. 

I’m opening this thread to centralize all the great ideas for the future roadmap. If you want to contribute, build a bridge, or just share what you are testing, this is the place!

#### 1. 📻 Offline Alerting & Mesh Networks (Meshtastic / LoRa)
* **The Goal:** When the ESP32 detects a drone, it shouldn't just print to serial—it should warn the neighborhood even if cell towers are down.
* **The Idea:** Integrating Batear with **Meshtastic**. We could use the ESP32's UART to send a simple text payload (e.g., `⚠️ Batear Node 1: Drone Detected!`) to a connected Meshtastic LoRa node.
* *Looking for:* Anyone familiar with the Meshtastic Serial API to help write a lightweight bridge or plugin.

#### 2. 🌬️ Environmental Noise & Hardware Mods
* **The Goal:** Reduce false positives from wind, cars, and lawnmowers. 
* **The Idea:** The Hackaday article mentioned using a "foam sock" or a physical acoustic amplifier (like a parabolic dish or "war tuba" cone) to improve directionality and block wind. 
* *Looking for:* 3D printing enthusiasts to design snap-on acoustic cones for the ICS-43434 mic, and folks who can test the thresholds outdoors in windy conditions.

#### 3. 🧠 TinyML / ESP-NN Upgrades
* **The Goal:** Move beyond static Goertzel frequency bins to recognize complex acoustic signatures.
* **The Idea:** Train a lightweight TensorFlow Lite Micro model that can classify drone rotors vs. background noise, while still fitting inside the ESP32-S3's SRAM and running locally.
* *Looking for:* ML engineers who want to help build an audio dataset (DJI, FPV, ambient noise) and train the first edge model.

---

**🤝 How to get involved:**
If you have an ESP32 and a micro-drone, please share your real-world test results below! Let me know what your `cal: active=N/6` output looks like when the drone is hovering vs. flying aggressively. 

Let's build a world where no one needs to fear the sky. 🦇 Drop your thoughts, links, and experiments below! 👇

**Comment by robinsandfort** [comment]
What a great project! 
I´m flying drones for conservation projects myself but I`m also involved 
in monitoring and protecting conservation areas like nationalparks and 
sensitive areas like nest sites of breeding raptors (like eagles) from illegal drone activity and disturbance. 
This would be a great use case with long range Lora connection (I already build and use a TTN network in the parc. 
Greetings from the woods,
Robin

**Reply by TN666** [reply]
Hi @robinsandfort, greetings from Taiwan!

Reading this absolutely made my day. Protecting nesting raptors from illegal drone disturbance is an incredible "Tech for Good" use case that I hadn't even fully considered. This is exactly why I open-sourced Batear!

Your timing couldn't be better regarding the long-range LoRa connection. Integrating LoRaWAN (TTN) and Meshtastic for remote edge alerts is actually the core goal for Milestone 2.

Since you already have a TTN gateway up and running in the national park, you would be the absolute perfect real-world beta tester for this.

Quick question regarding deploying in the woods: How much background audio (like wind howling through trees or loud bird calls) do you typically deal with? I imagine we might need to tweak the Goertzel/FFT thresholds or implement a dynamic noise-floor baseline so the system doesn't get confused by the forest itself!

So glad to have you here. Let's make the woods a safer place for the eagles!

**Reply by robinsandfort** [reply]
That really sounds great. 
We have thousands of hours of raw audio .wav recordings from the area. It should be easy to check or train for
background sounds. We mainly use audiomoth https://www.openacousticdevices.info/audiomoth with 48kHz.
We also use  https://github.com/cyberman54/ESP32-Paxcounter as an early warning system running on Heltec WiFi LoRa 32(V4), ESP32S3 + SX1262 LoRa Node.
Greetings from the woods, 
Robin

**Reply by TN666** [reply]
@robinsandfort This is fantastic, Robin. AudioMoth recordings are exactly the kind of high-quality dataset we need to stress-test the algorithm against real forest background noise.

Would it be possible for you to share a short snippet (maybe a few minutes) of a typical busy forest environment? I can run it through our Goertzel simulator locally to see if any natural sounds trigger false positives in the drone frequency bands (e.g., 400Hz, 800Hz, 1200Hz).

**Reply by robinsandfort** [reply]
@TN666 For sure. Where should I upload the files?

**Reply by TN666** [reply]
@robinsandfort  Thank you! Could you please upload the files to Google Drive (or any cloud service) and share the public link right here in this thread?

Sharing it publicly will allow other contributors in the community to use this valuable dataset to test and optimize the detection algorithms alongside us.

**Reply by robinsandfort** [reply]
Good morning @TN666 
here are some audio samples 
from the forest: https://www.dropbox.com/t/6OpgfnwLyo5mCX3H

We also published a giant dataset form urban to suburban from all over the world:

https://www.nature.com/articles/s41597-024-03611-7
https://osf.io/h285u/files/osfstorage

Greetings from the woods,
Robin

**Reply by TN666** [reply]
Hi @robinsandfort 

Thank you for the datasets and forest samples. They are exactly the "Golden Dataset" Batear needs to move from bench to field.

I’m currently benchmarking a 1024-point FFT implementation (esp-dsp) on the ESP32-S3 to improve frequency tracking. Your data will be the primary test for profiling ambient noise (wind, birds, etc.) and stress-testing our detection logic to minimize false positives.

Best,
TN

**Reply by robinsandfort** [reply]
Hi @TN666,
great! I also own a DJI Mini 4 Pro and A DJI Mavic Pro
and can also do sound recordings with different flight heights, directions and speeds.

Greeting from the woods,
Robin

**Reply by TN666** [reply]
Hi @robinsandfort, That would be incredible! Having specific recordings for the Mini 4 Pro and Mavic Pro at varying heights and speeds is exactly what I need for the next stage of validation. 

I have some solid benchmark data to share from the ESP32-S3:

```========== DSP Benchmark (FFT vs Goertzel) ==========
Sample rate : 16000 Hz
[Goertzel]  6 bins x 512 samples :   0.2 us  (0.000 ms)
[FFT-512]   512-pt complex          :  263.6 us  (0.264 ms)
[FFT-1024]  1024-pt complex      : 564.4 us  (0.564 ms)

--- Analysis ---
- Time to collect 1024 samples: 64,000 us (64 ms)
- Time to process 1024-pt FFT: 564.4 us
- CPU Overhead: ~0.88%
=====================================================
```

Even though the FFT is computationally heavier than my previous Goertzel implementation, it only consumes about 0.88% of the CPU cycle. This massive headroom allows Batear to analyze the entire spectrum in real-time without dropping any I2S frames.

With this efficiency, your proposed recordings will be the "Golden Dataset" for calibrating multi-harmonic detection. This is key to distinguishing DJI motor signatures from birds or wind noise in the woods.

Please go ahead with the recordings whenever you can. Can't wait to feed them into the new v1.2.0 engine!

Best,

TN

**Reply by robinsandfort** [reply]
Would it make sense to record with the same / similar microphone
setup? It is rainy and stormy here right now, we will have to wait. 
Greetings from the woods, Robin

**Reply by TN666** [reply]
Hi @robinsandfort 

Safety first—please do NOT fly in the rain! We definitely don't want to risk any damage to your DJI gear.

Regarding the microphone: Any setup is fine. Using different hardware (phone, recorder, etc.) actually helps me test the robustness of the FFT engine and ensures it's not "overfitted" to one specific device.

We can wait for a clear, dry day for the actual flight recordings. Stay dry and safe!

Best,
TN

**Reply by robinsandfort** [reply]
Hi @TN666,
I do have some audio files from an experiment.
I hung an audiorekorder (mini version of audiomoth) under 
my DJI mavic pro. 3-4 meters of rope between the drone and the recorder.
My aim was to record bats from above with the recorder hanging below. 
<img width="1621" height="673" alt="20240610_195705 WAV ethhc" src="https://github.com/user-attachments/assets/e7729de6-fc78-46ea-9095-24808cd273e2" />
The higher frequency signal are echo location calls from our local bats and you can see the low noise bed of the drone.
The recordings are quite noisy but maybe helpful?

**Reply by TN666** [reply]
Hi @robinsandfort 

This is incredible data! The separation you've captured between the Mavic Pro’s low-frequency motor noise and the high-frequency bat echolocation pulses is exactly the kind of 'edge case' we need to stress-test our DSP pipeline. 3-4 meters of rope was a very smart move to isolate structural vibrations.

To manage these high-fidelity recordings professionally without bloating our firmware source code, I’ve just launched a dedicated research repository:

👉 [batear-io/batear-datasets](https://github.com/batear-io/batear-datasets)

I have configured this new repo with Git LFS (Large File Storage) to handle raw .wav files. I would like to formally invite you to contribute these samples to our field-tests/mavic-series/ folder.

Your Mavic Pro recordings will become our 'Gold Standard' for:

Benchmark Testing: Validating our FFT binning accuracy in noisy environments.

Filter Optimization: Fine-tuning our High-Pass Filters to isolate biological signals from mechanical noise.

Downsampling Research: Testing how these signals behave when decimated to our 16kHz production target.

Would you be interested in uploading the raw files there? I can add you as a contributor to the organization so you can manage the data directly in the lab.

Looking forward to your high-res 'bat-detection' insights!"

**Comment by Clo363** [comment]
Here is a potentially life-saving feature idea from a real-world use case in Ukraine.

We are constantly attacked by Shahed-136 drones, and because alerts happen so frequently, people often cannot realistically run to shelters for every single warning. What would help enormously is distinguishing between a drone that is merely passing nearby and one that is actually flying toward your location.

Would it be possible to add directional tracking / trajectory estimation for Shahed-type acoustic signatures?

The drone has a very distinctive sound — similar to a very loud 2-stroke trimmer engine. In practice, we can often hear it clearly at roughly 3 km distance or closer.

Some ideas that might make this feasible:

* A microphone array could estimate bearing and track movement over time
* A distributed mesh of low-cost LoRa-connected microphones could triangulate direction and trajectory
* Since the drone speed is relatively constant (~185 km/h except during terminal dive), prediction algorithms may work well
* The target is usually significantly above ambient noise level
* The sound signature is easy to emulate for testing using a 2-stroke trimmer at high RPM

Even approximate “moving toward you / moving away / passing sideways” classification could greatly improve civilian response decisions and reduce alert fatigue.

**Reply by TN666** [reply]
@Clo363 Thank you for this — the real-world context is exactly the kind of feedback that shapes priorities.

Of the three approaches you outlined, single-node Doppler trend analysis looks like the most realistic path on existing hardware: track the dominant harmonic's frequency drift over a few seconds and classify *approaching / receding / passing*. It directly addresses the alert-fatigue problem without adding microphones or relying on mesh timing.

Microphone arrays would push BOM and enclosure complexity beyond batear's low-cost positioning, and LoRa's packet latency (hundreds of ms) is too coarse for TDOA triangulation in real time — so those two are unlikely for the mainline firmware, though mesh coordination is being tracked separately in #51.

I'll prototype the Doppler approach offline first and open a tracking issue once I have something to validate. If you have any Shahed recordings you can share, even phone-quality, that would help enormously. Stay safe.

**Reply by Clo363** [reply]
Thanks for the reply! Take into account that drones are sent in multiples, so there are several at once in the air. Regarding microphones array wont a local 4 analogue microphones do the trick? Spread apart and have some auto position calibration on a clap. If its a LoRa mesh a GPS for expre prescice time would be needed, not to rely on signal latancy. Register distinct sound + exact time + moving to/away could allow real positioning with enough nodes (it seems military uses something like this, but they dont share exact location with civil population).
